### PR TITLE
Week 1 day 2

### PR DESCRIPTION
**Pull request by emperorliu**

### CODE REVIEW

File: lib/iteration.rb [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+def factors(num)
+  (1..num).select { |n| num % n == 0 }
+end
+
+class Array
+  def bubble_sort!(&prc)
+    sorted = false
+    until sorted
+      sorted = true
+      (0..self.length - 2).each do |idx|
```
Comment (sophylee) [inline lib/iteration.rb:34]: Why `length - 2`?

File: lib/iteration.rb [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    p yield *args
+  end
+end
+
+def factors(num)
+  (1..num).select { |n| num % n == 0 }
+end
+
+class Array
+  def bubble_sort!(&prc)
```
Comment (sophylee) [inline lib/iteration.rb:30]: As a sanity check, could you explain in plain english how bubble sort works?

**Comment by sophylee** [comment]
LGTM

File: lib/iteration.rb (hunk 2 above)
Comment (emperorliu) [inline lib/iteration.rb:30]: Yes. You sort by comparing the current element to the one after, and keep going until the end of the array. Then you repeat. The largest number would naturally get sorted to the end the first round.

File: lib/iteration.rb (hunk 1 above)
Comment (emperorliu) [inline lib/iteration.rb:34]: Because in an array like [ 1, 3, 4, 5, 6 ] which has 5 elements, I only want to sort from position 0-3, which is basically the numbers 1, 3, 4, 5. Once I compare 5 with 6, I don't need to move forward anymore because I don't have any thing to compare the last element with.

File: lib/iteration.rb (hunk 2 above)
Comment (sophylee) [inline lib/iteration.rb:30]: Good

File: lib/iteration.rb (hunk 1 above)
Comment (sophylee) [inline lib/iteration.rb:34]: Good
